### PR TITLE
Cleanup project files

### PR DESCRIPTION
**Pull request by js6pak**

### CODE REVIEW

File: .github/workflows/codeql-analysis.yml
Comment (IEvangelist) [inline .github/workflows/codeql-analysis.yml:1]: I'm good with these changes.

File: .github/workflows/publish-nuget.yml
Comment (IEvangelist) [inline .github/workflows/publish-nuget.yml:1]: I prefer having the publishing to NuGet being in its own workflow, let's leave this as is.

File: .github/workflows/pr-validation.yml
Comment (IEvangelist) [inline .github/workflows/pr-validation.yml:1]: Yeah, I'm okay with this being conjoined with the `build-and-test.yml` that makes sense and is fine.

File: .github/workflows/build-and-test.yml [hunk 1]
```diff
@@ -2,48 +2,77 @@ name: build-and-test
...
+
+      - name: Test
+        run: dotnet test --configuration Release --no-restore
+
+      - name: Publish
+        if: matrix.os == 'ubuntu-latest' && github.ref_type == 'tag'
+        run: |
+          dotnet nuget push --skip-duplicate artifacts/package/release/*.nupkg \
+            --source "https://api.nuget.org/v3/index.json" --api-key "${{ secrets.NUGET_API_KEY }}"
+
```
Comment (IEvangelist) [inline .github/workflows/build-and-test.yml]: Let's leave the publish in its own workflow.

```suggestion
```

File: .github/workflows/build-and-test.yml [hunk 2]
```diff
@@ -2,48 +2,77 @@ name: build-and-test
...
+  test-samples:
```
Comment (IEvangelist) [inline .github/workflows/build-and-test.yml:39]: I love the testing of the sample apps, that was a goal I hadn't gotten around to yet, thank you. But, I really don't want these to run with every `push` and `pull_request`, let's move these into their own workflow and perhaps make them a requirement to pass before publishing to NuGet, yeah?

File: .editorconfig [hunk 3]
```diff
@@ -6,9 +6,14 @@ root = true
...
 # All files 
 [*] 
-indent_style = space 
+indent_style = space
+insert_final_newline = true
 file_header_template = Copyright (c) David Pine. All rights reserved.\nLicensed under the MIT License.
 
+[*.{xml,*proj,props,targets,ruleset,config,nuspec,resx}]
+indent_size = 4
+ij_xml_space_inside_empty_tag = true
```
Comment (IEvangelist) [inline .editorconfig:15]: What is `ij_xml_space_inside_empty_tag` I'm not familiar with that one?

File: Directory.Build.props
Comment (IEvangelist) [inline Directory.Build.props:1]: It makes a lot of sense to encapsulate all of these shared bits into the build, thank you. But I think there is still a bit of work to be done here. For example, not everything is accounted for. I'll call out a few discrepancies in the individual projects.

File: Directory.Build.props [hunk 4]
```diff
@@ -1,31 +1,34 @@
...
+        <EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
+        <ReportAnalyzer>true</ReportAnalyzer>
 
-  <PropertyGroup>
-    <ImplicitUsings>enable</ImplicitUsings>
-    <Nullable>enable</Nullable>
-  </PropertyGroup>
+        <ContinuousIntegrationBuild Condition="'$(GITHUB_ACTIONS)' == 'true'">true</ContinuousIntegrationBuild>
+        <!-- TODO re-enable -->
+        <!-- <TreatWarningsAsErrors Condition="'$(ContinuousIntegrationBuild)' == 'true' or '$(Configuration)' == 'Release'">true</TreatWarningsAsErrors> -->
```
Comment (IEvangelist) [inline Directory.Build.props:21]: C'mon now, nice try 🤓. Let's clean up all the warnings before we merge this, please re-enable. And let's make it unconditional. We shall always treat warnings as errors to help us avoid having warnings...


```suggestion
        <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
```

File: Directory.Build.props [hunk 5]
```diff
@@ -1,31 +1,34 @@
...
 
-  <PropertyGroup>
-    <IsPackable>true</IsPackable>
-  </PropertyGroup>
+        <Authors>David Pine</Authors>
+        <Copyright>2022 David Pine</Copyright>
```
Comment (IEvangelist) [inline Directory.Build.props]: This doesn't match the originals:
```suggestion
        <Copyright>© 2022-20224 David Pine</Copyright>
```

File: Directory.Build.props [hunk 6]
```diff
@@ -1,31 +1,34 @@
...
-  </PropertyGroup>
-
-  <PropertyGroup>
-    <SourceLinkCreate>true</SourceLinkCreate>
-    <SourceLinkOriginUrl>https://github.com/IEvangelist/dotnet-github-actions-sdk</SourceLinkOriginUrl>
-  </PropertyGroup>
-</Project>
\ No newline at end of file
+        <EmbedUntrackedSources>true</EmbedUntrackedSources>
+    </PropertyGroup>
```
Comment (IEvangelist) [inline Directory.Build.props:33]: Why was the source link bits removed?

```suggestion
    </PropertyGroup>
    
    <PropertyGroup>
        <SourceLinkCreate>true</SourceLinkCreate>
        <SourceLinkOriginUrl>https://github.com/IEvangelist/dotnet-github-actions-sdk</SourceLinkOriginUrl>
  </PropertyGroup>
```

File: global.json
Comment (IEvangelist) [inline global.json:1]: Grumble... I'm not a huge fan of this. I was just told by some of my peers on the .NET team here at Microsoft that this is something that should be discouraged. Why do you prefer it?

File: samples/Actions.Core.Sample/Dockerfile
Comment (IEvangelist) [inline samples/Actions.Core.Sample/Dockerfile:1]: I'm thinking that perhaps we should just delete the _Dockerfile_ altogether, and instead just do `dotnet publish` from a CI that publishes the app as a container image. Similar to what I've done here: https://github.com/IEvangelist/profanity-filter/blob/main/.github/workflows/release.yml 

That might clean things up a bit and make it faster to test, thoughts?

File: src/Actions.Core/Actions.Core.csproj [hunk 7]
```diff
@@ -1,77 +1,11 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
+    <PropertyGroup>
+        <TargetFrameworks>$(DefaultTargetFrameworks)</TargetFrameworks>
 
-  <PropertyGroup>
-    <TargetFrameworks>$(DefaultTargetFrameworks)</TargetFrameworks>
-    <ImplicitUsings>enable</ImplicitUsings>
-    <Nullable>enable</Nullable>
-    <Description>This is an unofficial .NET SDK for GitHub Actions workflows (based on @actions/core).</Description>
```
Comment (IEvangelist) [inline src/Actions.Core/Actions.Core.csproj:7]: The following bits are missing that need to be added back to each project:

- `NeutralLanguage`
- `Version`
- `VersionSuffix`
- `FileVersion`
- `ClientVersion`
- `AssemblyName`
- `PackageId`
- `RootNamespace`
- `Optimize`
- `PackageLicenseFile`

And for testing, `InternalsVisibleTo`.

File: src/Actions.Core/build/GitHub.Actions.Core.props
Comment (IEvangelist) [inline src/Actions.Core/build/GitHub.Actions.Core.props:1]: Why are these deleted? They're required for the apps running as a container to function correctly. See PR #1.

**Review by IEvangelist — CHANGES_REQUESTED** [review]
I really love where this is heading, thank you so much for this! Much appreciated. I have some feedback, but once that's addressed let's get this merged.

File: .github/workflows/build-and-test.yml (hunk 2 above)
Comment (js6pak) [reply .github/workflows/build-and-test.yml:39]: > But, I really don't want these to run with every push and pull_request

Why not? Open source github actions are free and it runs pretty quick so imo it's better to know if you've broken them in any way right away.

File: .editorconfig (hunk 3 above)
Comment (js6pak) [reply .editorconfig:15]: IntelliJ/Rider option which puts a space before `/>` in xml, annoyingly it's off by default, but it's the standard in .NET (and you used it here aswell)
image

File: Directory.Build.props (hunk 4 above)
Comment (js6pak) [reply Directory.Build.props:21]: Most of the new warnings come from me setting `AnalysisMode` to `recommended`, I was hoping you would go over them and fix/disable ones you don't think are necessary.

> And let's make it unconditional

The idea is it should compile with warnings in Debug mode so it doesn't annoy you during development and prototyping.

File: Directory.Build.props (hunk 5 above)
Comment (js6pak) [reply Directory.Build.props]: I've changed this on purpose to conform with the [REUSE/SPDX convention](https://reuse.software/faq/#copyright-symbol).
Feel free to revert.

File: Directory.Build.props (hunk 6 above)
Comment (js6pak) [reply Directory.Build.props:33]: Can you point me to where you got `SourceLinkCreate`/`SourceLinkOriginUrl` from? I don't think those actually do anything, and they are not documented anywhere.
Referencing `Microsoft.SourceLink.GitHub` and setting `RepositoryUrl` should be enough.

File: global.json
Comment (js6pak) [reply global.json:1]: .NET SDK major versions introduce major breaking changes...
Ideally every contributor and CI would have the exact same SDK version, but there's no good way to achieve that, so we should at least ensure we are running the same major version.

> I was just told by some of my peers on the .NET team here at Microsoft that this is something that should be discouraged

Why?

File: samples/Actions.Core.Sample/Dockerfile
Comment (js6pak) [reply samples/Actions.Core.Sample/Dockerfile:1]: Initially I wanted to keep using the `using: Dockerfile` inside `action.yml` but before I realized there is no way(?) to set the build context to repository root I already had a `Dockerfile` working, so I didn't look into setting up 
 `Microsoft.NET.Build.Containers`. Last time I tried it with CLI apps it didn't work though so I don't know.

File: src/Actions.Core/Actions.Core.csproj (hunk 7 above)
Comment (js6pak) [reply src/Actions.Core/Actions.Core.csproj:7]: > NeutralLanguage

Does that even do anything in a library?

> Version
	VersionSuffix
	FileVersion
	ClientVersion

These are all redundant due to MinVer which you were already using.

> AssemblyName
	RootNamespace
	Optimize

Those were already set to the default values.

> PackageLicenseFile

You use standard MIT, so I replaced it with `PackageLicenseExpression` (+ `Copyright`)

> PackageId

https://github.com/js6pak/dotnet-github-actions-sdk/blob/7618722d40a7d7ab1b4ba843e1d586bc6e7ab02d/src/Directory.Build.targets#L5

> InternalsVisibleTo

https://github.com/js6pak/dotnet-github-actions-sdk/blob/7618722d40a7d7ab1b4ba843e1d586bc6e7ab02d/src/Directory.Build.targets#L14

File: src/Actions.Core/build/GitHub.Actions.Core.props
Comment (js6pak) [reply src/Actions.Core/build/GitHub.Actions.Core.props:1]: Deduplicated it because it was all the same anyway
https://github.com/js6pak/dotnet-github-actions-sdk/blob/7618722d40a7d7ab1b4ba843e1d586bc6e7ab02d/src/Directory.Build.props#L11

File: .github/workflows/publish-nuget.yml
Comment (js6pak) [reply .github/workflows/publish-nuget.yml:1]: You'd end up with 90% of the file being duplicated.

File: .github/workflows/codeql-analysis.yml
Comment (js6pak) [reply .github/workflows/codeql-analysis.yml:1]: Does CodeQL even have any actual value for C#?

Comment (IEvangelist) [reply .github/workflows/codeql-analysis.yml:1]: Oh yes, very much so - it's actually caught several things across a few of my different open-source projects that was otherwise unknown to me to be a risk. I learn from it, often. For example, consider: https://codeql.github.com/docs/codeql-language-guides/analyzing-data-flow-in-csharp - which is of course, just a part of it.

File: .github/workflows/publish-nuget.yml
Comment (IEvangelist) [reply .github/workflows/publish-nuget.yml:1]: I see the concern, but what I'm saying is we can shortcut the call to the testing bits if need be, encapsulate them in their own workflow so that can then be called upon in both consuming workflows to help eliminate the duplication concern. See this example: https://stackoverflow.com/a/64733705/2410379.

File: .github/workflows/build-and-test.yml (hunk 2 above)
Comment (IEvangelist) [reply .github/workflows/build-and-test.yml:39]: How long will these take to run? I've seen it take up to two full minutes for building a single container, and running it. This is proposing that we'd have two built, with imaginably more for each test project sample we add. It's not even happening in parallel, so each time this runs we could be seeing four minute runs on a PR, that's a long time IMO (basically add 2 minutes for every project sample 🙈). We'll write more unit tests.

File: .editorconfig (hunk 3 above)
Comment (IEvangelist) [reply .editorconfig:15]: I'm good with it, was just curious. That's my preference too.

File: Directory.Build.props (hunk 4 above)
Comment (IEvangelist) [reply Directory.Build.props:21]: What are some of the warnings with the recommended analysis mode? How much will need to change, will I find myself updating the `.editorconfig` more?

File: Directory.Build.props (hunk 6 above)
Comment (IEvangelist) [reply Directory.Build.props:33]: It's an internal implementation detail: https://github.com/ctaggart/SourceLink/issues/327#issuecomment-380303070

File: global.json
Comment (IEvangelist) [reply global.json:1]: > Why?

"The versions for the runtime packs are all hardcoded into the SDK. Since the "latest" SDK on your machine might be older than one of the down-level runtimes (e.g., .NET 7 SDK, .NET 6 runtime with recent patches), the patch versions can be dropped when publishing, since the SDK just uses the hardcoded versions."

**Comment by IEvangelist** [comment]
BTW: I want this to be clear, I really love your contributions on this @js6pak - this is going to be awesome, thank you!

**Comment by IEvangelist** [comment]
@all-contributors plepase add @js6pak for code and tests

File: Directory.Build.props [hunk 8]
```diff
@@ -21,7 +21,7 @@
         <!-- <TreatWarningsAsErrors Condition="'$(ContinuousIntegrationBuild)' == 'true' or '$(Configuration)' == 'Release'">true</TreatWarningsAsErrors> -->
 
         <Authors>David Pine</Authors>
-        <Copyright>2022 David Pine</Copyright>
+        <Copyright>© 2022-20224 David Pine</Copyright>
```
Comment (js6pak) [inline Directory.Build.props]: > 20224

File: .github/workflows/build-and-test.yml (hunk 2 above)
Comment (js6pak) [reply .github/workflows/build-and-test.yml:60]: Currently, it takes [46s total](https://github.com/IEvangelist/dotnet-github-actions-sdk/actions/runs/8051843926/job/21999555147?pr=6) to build and run 2 samples, and it might be even faster with `Microsoft.NET.Build.Containers`. It runs in parallel with normal build, so it literally doesn't matter.

> We'll write more unit tests.

Will unit tests be able to catch changes made by github?

File: global.json
Comment (js6pak) [reply global.json:1]: That's a non issue because we only lock the major/minor version by setting `rollForward` to `latestFeature` (`8.0.xxx`).

File: .github/workflows/build-and-test.yml (hunk 2 above)
Comment (js6pak) [reply .github/workflows/build-and-test.yml:39]: Currently, it takes [46s total](https://github.com/IEvangelist/dotnet-github-actions-sdk/actions/runs/8051843926/job/21999555147?pr=6) to build and run 2 samples, and it might be even faster with `Microsoft.NET.Build.Containers`. It runs in parallel with normal build, so it literally doesn't matter.

> We'll write more unit tests.

Will unit tests be able to catch changes made by github?

Comment (IEvangelist) [reply .github/workflows/build-and-test.yml:39]: I suppose that's not the end of the world, but I reserve the right to pull this out if I get annoyed with it later. 😛

**Review by IEvangelist — APPROVED** [review]
Looks good to me, we'll :shipit: - thank you 🙏
